cnb_index.go

```diff
@@ -0,0 +1,381 @@
+func (h *CNBIndex) SaveDir() error {
+	layoutPath := filepath.Join(h.XdgPath, MakeFileSafeName(h.RepoName)) // FIXME: do we create an OCI-layout compatible directory structure?
+	var (
+		path layout.Path
+		err  error
+	)
+
+	if _, err = os.Stat(layoutPath); !os.IsNotExist(err) {
+		// We need to always init an empty index when saving
+		if err = os.RemoveAll(layoutPath); err != nil {
+			return err
+		}
+	}
+
+	indexType, err := h.ImageIndex.MediaType()
+	if err != nil {
+		return err
+	}
+	if path, err = newEmptyLayoutPath(indexType, layoutPath); err != nil {
+		return err
+	}
+
+	var errs SaveError
+	index, err := h.ImageIndex.IndexManifest()
+	if err != nil {
+		return err
+	}
+	for _, desc := range index.Manifests {
+		appendManifest(desc, path, &errs)
+	}
+	if len(errs.Errors) != 0 {
+		return errs
+	}
+	return nil
+}
+
+func appendManifest(desc v1.Descriptor, path layout.Path, errs *SaveError) {
+	if err := path.RemoveDescriptors(match.Digests(desc.Digest)); err != nil {
+		errs.Errors = append(errs.Errors, SaveDiagnostic{
+			Cause: err,
+		})
+	}
+	if err := path.AppendDescriptor(desc); err != nil {
+		errs.Errors = append(errs.Errors, SaveDiagnostic{
+			Cause: err,
+		})
+	}
+}
+
+func newEmptyLayoutPath(indexType types.MediaType, path string) (layout.Path, error) {
+	if indexType == types.OCIImageIndex {
+		return layout.Write(path, empty.Index)
+	}
+	return layout.Write(path, NewEmptyDockerIndex())
+}
+
+// Push Publishes ImageIndex to the registry assuming every image it referes exists in registry.
+//
+// It will only push the IndexManifest to registry.
+func (h *CNBIndex) Push(ops ...func(*IndexPushOptions) error) error {
+	var pushOps = &IndexPushOptions{}
+	for _, op := range ops {
+		if err := op(pushOps); err != nil {
+			return err
+		}
+	}
+
+	if pushOps.Format != "" {
+		if !pushOps.Format.IsIndex() {
+			return ErrUnknownMediaType(pushOps.Format)
+		}
+		existingType, err := h.ImageIndex.MediaType()
+		if err != nil {
+			return err
+		}
+		if pushOps.Format != existingType {
+			h.ImageIndex = mutate.IndexMediaType(h.ImageIndex, pushOps.Format)
+		}
+	}
+
+	ref, err := name.ParseReference(
+		h.RepoName,
+		name.WeakValidation,
+		name.Insecure,
+	)
+	if err != nil {
+		return err
+	}
+
+	indexManifest, err := getIndexManifest(h.ImageIndex)
+	if err != nil {
+		return err
+	}
+
+	var taggableIndex = NewTaggableIndex(indexManifest)
+	multiWriteTagables := map[name.Reference]remote.Taggable{
+		ref: taggableIndex,
+	}
+	for _, tag := range pushOps.Tags {
+		multiWriteTagables[ref.Context().Tag(tag)] = taggableIndex
+	}
+
+	// FIXME: this will only push the index manifest, assuming that all the images it refers to exist in the registry
+	err = remote.MultiWrite(
+		multiWriteTagables,
+		remote.WithAuthFromKeychain(h.KeyChain),
+		remote.WithTransport(GetTransport(pushOps.Insecure)),
+	)
+	if err != nil {
+		return err
+	}
+
+	if pushOps.Purge {
+		return h.DeleteDir()
+	}
+	return h.SaveDir()
```

Comment:
isn't the `--purge` option used to delete local imageIndex after the pushing is successful?

Reply:
Yeah, but we return on line 338 so if we purge we never get to line 340